src/java.base/share/classes/java/util/HashMap.java

```diff
@@ -2545,4 +2545,55 @@ else if ((xpl = xp.left) == x) {
         }
     }
 
+    /**
+     * Calculate initial capacity for HashMap based classes, from expected size and default load factor (0.75).
+     *
+     * <p>This function is designed to return equal results than {@code (int)Math.ceil(expectedSize / 0.75)} when expectedSize&gt;=0.
+     * When expectedSize&lt;0, result of this function is meaningless.
+     *
+     * <p>Implementation logic of this function is:
+     *
+     * <p>1. When expectedSize is a small enough positive number,
+     * {@code (expectedSize + (expectedSize + 2) / 3)} always equals to {@code (int)Math.ceil(expectedSize / 0.75)}, as
+     * {@code (int)Math.ceil(expectedSize / 0.75)} equals to {@code (int)Math.ceil(expectedSize + expectedSize / 3.0)},
+     * thus equals to {@code expectedSize + (int)Math.ceil(expectedSize / 3.0)},
+     * thus equals to {@code (expectedSize + (expectedSize + 2) / 3)}.
+     *
+     * <p>2. When expectedSize is a big enough positive number (at least {@code Integer.MAX_VALUE / 4 * 3 + 3}, or say, 1610612736),
+     * {@code (expectedSize + (expectedSize + 2) / 3)} would overflow, in this turn we just return {@code Integer.MAX_VALUE}
+     *
+     * <p>3. When expectedSize is 0, return 0.
+     *
+     * <p>4. User must never let expectedSize be negative.Otherwise, return any number &lt; 0.
+     *
+     * @param expectedSize expected size
+     * @return initial capacity for HashMap based classes.
+     * @since 19
+     */
+    static int calculateHashMapCapacity(int expectedSize) {
+        if (expectedSize >= Integer.MAX_VALUE / 4 * 3 + 3) {
+            return Integer.MAX_VALUE;
+        }
+        if (expectedSize > 0) {
+            return (expectedSize + (expectedSize + 2) / 3);
+        }
+        return expectedSize;
+    }
+
+    /**
+     * Creates a new, empty HashMap with an initial table size
```

Comment:
You probably wanna allow for a non-NEW instance for the corner case where the given size is 0 - no elements.

Reply:
> You probably wanna allow for a non-NEW instance for the corner case where the given size is 0 - no elements.

@ChrisHegarty I guess we shouldn't.

I want to make it 100% equals to `new HashMap()` constructor, thus migrate all usecases.

So if we apply this, and when the original usage use this map object as a lock, or put some elements after the call(sometimes people cannot decide if they would really put elements in this map), bad things would happen.

Besides, there already a function` Map.of()` for such functionality, so programmer should use that instead if they really want a shared empty map.

Reply:
hash maps are modifiable. empty instances can be changed, and returning such shared instances are inherently unsafe.

Reply:
Yeah, sorry. Ignore my comment, I was wrong. I completely agree with your reasoning.